service/history/api/respondworkflowtaskcompleted/api.go

```diff
@@ -252,7 +254,10 @@ func (handler *WorkflowTaskCompletedHandler) Invoke(
 	}()
 
 	// It's an error if the workflow has used versioning in the past but this task has no versioning info.
-	if ms.GetMostRecentWorkerVersionStamp().GetUseVersioning() && !request.GetWorkerVersionStamp().GetUseVersioning() {
+	if ms.GetMostRecentWorkerVersionStamp().GetUseVersioning() &&
+		//nolint:staticcheck // SA1019 deprecated stamp will clean up later
+		!request.GetWorkerVersionStamp().GetUseVersioning() &&
+		request.GetDeploymentOptions().GetWorkerVersioningMode() != enumspb.WORKER_VERSIONING_MODE_VERSIONED {
 		// Mutable state wasn't changed yet and doesn't have to be cleared.
 		releaseLeaseWithError = false
 		return nil, serviceerror.NewInvalidArgument("Workflow using versioning must continue to use versioning.")
```

Comment:
how does "ramp to unversioned" work with this? If someone uses versioning in a version stamp, they can't ramp to unversioned? (It makes sense to me that v1/v2 can't ramp to unversioned, but v3.1 should be able to, right?)

Reply:
yeah they could not in v1-2. now that I exclude v3.1 from the condition it should be possible for v3.1 even if stamp is not sent.